libs/MeshKernel/src/Polygon.cpp

```diff
@@ -254,52 +254,97 @@ std::vector<double> meshkernel::Polygon::EdgeLengths() const
     return edgeLengths;
 }
 
+// Refines the section from node specified by index to the next node. The first node is added, the next node is not
+void RefineSegment(std::vector<meshkernel::Point>& refinedPolygon,
+                   const std::vector<meshkernel::Point>::const_iterator& nodeIterator,
+                   const double refinementDistance,
+                   const meshkernel::Projection projection)
+{
+    // Line segment starting point.
+    const auto& n0 = *nodeIterator;
+    const auto& n1 = *std::next(nodeIterator);
+
+    refinedPolygon.push_back(n0);
+
+    const double segmentLength = ComputeDistance(n0, n1, projection);
+
+    int n = std::lround(segmentLength / refinementDistance);
+    const double refinedLength = n * refinementDistance;
+    if (refinedLength > segmentLength ||
+        meshkernel::IsEqual(refinedLength, segmentLength, meshkernel::constants::geometric::refinementTolerance))
+        --n;
+
+    // Refined segment step size.
+    const meshkernel::Point delta = (n1 - n0) * refinementDistance / segmentLength;
+    for (auto i = 1; i <= n; ++i)
+    {
+        refinedPolygon.push_back(n0 + i * delta);
+    }
+}
+
 std::vector<meshkernel::Point> meshkernel::Polygon::Refine(const size_t startIndex, const size_t endIndex, const double refinementDistance) const
 {
+    if (startIndex == endIndex)
+    {
+        return m_nodes;
+    }
 
-    if (startIndex > endIndex || endIndex >= m_nodes.size())
+    if (startIndex >= m_nodes.size() || endIndex >= m_nodes.size())
     {
         throw ConstraintError("The indices are not valid: {}, {}.", startIndex, endIndex);
     }
 
     std::vector<Point> refinedPolygon;
-    refinedPolygon.reserve(m_nodes.size());
+    const auto iStart = static_cast<std::vector<Point>::difference_type>(startIndex);
+    const auto iEnd = static_cast<std::vector<Point>::difference_type>(endIndex);
+    const auto from = std::next(m_nodes.begin(), iStart);
+    const auto to = std::next(m_nodes.begin(), iEnd);
 
-    // Add nodes before the section to be refined
-    for (size_t i = 0; i <= startIndex; ++i)
+    auto computeDistance = [this](auto l, auto p)
```

Comment:
No need to capture this. Maybe a bit over the top but you can write  `auto computeDistance = [&projection = std::as_const(m_projection)](auto l, auto p)` (requires including utility).

Working example:

```
#include <vector>
#include <iostream>
#include <utility>
#include <map>
#include <string>

enum class Projection {
    Cartesian,
    Spherical,
    SphericalAccurate
};

std::map<Projection, std::string> ProjectionMap = {
    {Projection::Cartesian, "Cartesian"},
    {Projection::Spherical, "Spherical"},
    {Projection::SphericalAccurate, "Spherical accurate"}
};

class Foo
{
public:
    Foo(Projection projection) : m_projection(projection) {}

    std::string bar()  {
        return [&projection = std::as_const(m_projection)](){ 
            //projection = Projection::Spherical; // fails, const ref captured
            return ProjectionMap.at(projection); 
        }();
    }
private:
    Projection m_projection;
};

int main()
{
    Foo foo(Projection::Cartesian);
    std::cout <<  foo.bar() << '\n';
    return 0;
}
```

Reply:
done